pkg/admissioncontroller/webhooks/admission/internaldomainsecret/admission.go

```diff
@@ -0,0 +1,169 @@
+	}
+
+	switch request.Operation {
+	case admissionv1.Create:
+		exists, err := h.internalDomainSecretExists(ctx)
+		if err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+		if exists {
+			return admission.Denied("cannot create internal domain secret because there can be only one secret with the 'internal-domain' secret role")
+		}
+
+		secret := &corev1.Secret{}
+		if err := h.decoder.Decode(request, secret); err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+		if _, _, _, _, err := common.GetDomainInfoFromAnnotations(secret.Annotations); err != nil {
+			return admission.Errored(http.StatusUnprocessableEntity, err)
+		}
+
+		return acadmission.Allowed("internal domain secret is valid")
+
+	case admissionv1.Update:
+		secret, oldSecret := &corev1.Secret{}, &corev1.Secret{}
+		if err := h.decoder.Decode(request, secret); err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+		if err := h.decoder.DecodeRaw(request.OldObject, oldSecret); err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+
+		_, oldDomain, _, _, err := common.GetDomainInfoFromAnnotations(oldSecret.Annotations)
+		if err != nil {
+			return admission.Errored(http.StatusUnprocessableEntity, err)
+		}
+		_, newDomain, _, _, err := common.GetDomainInfoFromAnnotations(secret.Annotations)
+		if err != nil {
+			return admission.Errored(http.StatusUnprocessableEntity, err)
+		}
+
+		if oldDomain != newDomain {
+			atLeastOneShoot, err := h.atLeastOneShootExists(ctx)
+			if err != nil {
+				return admission.Errored(http.StatusInternalServerError, err)
+			}
+			if atLeastOneShoot {
+				return admission.Denied("cannot change domain because there are still shoots left in the system")
+			}
+		}
+
+		return acadmission.Allowed("domain didn't change or no shoot exists")
+
+	case admissionv1.Delete:
+		atLeastOneShoot, err := h.atLeastOneShootExists(ctx)
+		if err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+		if atLeastOneShoot {
+			return admission.Denied("cannot delete internal domain secret because there are still shoots left in the system")
+		}
+		return acadmission.Allowed("no shoot exists")
+
+	default:
+		return acadmission.Allowed("unknown operation")
+	}
+}
+
+func (h *handler) atLeastOneShootExists(ctx context.Context) (bool, error) {
+	shoots := &metav1.PartialObjectMetadataList{}
+	shoots.SetGroupVersionKind(gardencorev1beta1.SchemeGroupVersion.WithKind("ShootList"))
+
+	if err := h.apiReader.List(ctx, shoots, client.Limit(1)); err != nil {
+		return false, err
+	}
+
+	return len(shoots.Items) > 0, nil
+}
+
+func (h *handler) internalDomainSecretExists(ctx context.Context) (bool, error) {
+	secrets := &metav1.PartialObjectMetadataList{}
+	secrets.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("SecretList"))
+
+	if err := h.apiReader.List(
```

Comment:
As you are listing secrets here do you also need to adapt charts/gardener/controlplane/charts/application/templates/rbac-system.yaml similiar to https://github.com/gardener/gardener/pull/3788 ?

Reply:
Yep, thanks for catching this! Done.

Reply:
Does a list call always require the watch verb as well?

Reply:
No, but generally, if you allow LIST then WATCH and GET can also be allowed (they are quasi a subset of LIST).